File: examples/canonical-auth/java/src/main/java/dev/tenantcore/auth/routes/GroupController.java
package dev.tenantcore.auth.routes;

import dev.tenantcore.auth.auth.AuthContext;
import dev.tenantcore.auth.domain.Group;
import dev.tenantcore.auth.domain.InMemoryStore;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;
import java.util.Map;

@RestController
@RequestMapping("/groups")
public class GroupController {

    private final InMemoryStore store;

    public GroupController(InMemoryStore store) {
        this.store = store;
    }

    private AuthContext getAuth() {
        return (AuthContext) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
    }

    @GetMapping
    @PreAuthorize("hasAuthority('iam:group:read')")
    public ResponseEntity<Collection<Group>> listGroups() {
        return ResponseEntity.ok(store.getAllGroups(getAuth().tenantId()));
    }

    @PostMapping
    @PreAuthorize("hasAuthority('iam:group:create')")
    public ResponseEntity<?> createGroup(@RequestBody Group group) {
        Group created = store.createGroup(getAuth().tenantId(), group.slug(), group.name());
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }

    @PostMapping("/{id}/permissions")
    @PreAuthorize("hasAuthority('iam:group:assign_permission')")
    public ResponseEntity<?> assignPermission(@PathVariable String id, @RequestBody Map<String, String> body) {
        boolean assigned = store.assignPermissionToGroup(id, body.get("permission"), getAuth().tenantId());
        if (!assigned) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
        return ResponseEntity.ok().build();
    }

    @PostMapping("/{id}/users")
    @PreAuthorize("hasAuthority('iam:group:assign_user')")
    public ResponseEntity<?> assignUser(@PathVariable String id, @RequestBody Map<String, String> body) {
        boolean assigned = store.assignUserToGroup(id, body.get("user_id"), getAuth().tenantId());
        if (!assigned) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
        return ResponseEntity.ok().build();
    }
}

File: examples/canonical-auth/elixir/lib/tenantcore_auth/controllers/group_controller.ex
defmodule TenantcoreAuth.GroupController do
  use TenantcoreAuth, :controller

  alias TenantcoreAuth.Domain.InMemoryStore

  plug TenantcoreAuth.Auth.RbacPlug, "iam:group:read" when action in [:index]
  plug TenantcoreAuth.Auth.RbacPlug, "iam:group:create" when action in [:create]
  plug TenantcoreAuth.Auth.RbacPlug, "iam:group:assign_permission" when action in [:assign_permission]
  plug TenantcoreAuth.Auth.RbacPlug, "iam:group:assign_user" when action in [:assign_user]

  def index(%Plug.Conn{assigns: %{auth: auth}} = conn, _params) do
    groups =
      InMemoryStore.list_groups(auth.tenant_id)
      |> Enum.map(fn group -> %{id: group.id, tenant_id: group.tenant_id, slug: group.slug, name: group.name} end)

    json(conn, %{groups: groups})
  end

  def create(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"slug" => slug, "name" => name} = params) do
    permission_names = Map.get(params, "permission_names", [])
    known_permissions = InMemoryStore.list_permissions() |> Enum.map(& &1.name) |> MapSet.new()
    unknown_permissions = Enum.reject(permission_names, &MapSet.member?(known_permissions, &1))

    if unknown_permissions == [] do
      group = InMemoryStore.create_group(auth.tenant_id, slug, name, permission_names)
      conn |> put_status(:created) |> json(%{group: %{id: group.id, tenant_id: group.tenant_id, slug: group.slug, name: group.name}})
    else
      conn |> put_status(:bad_request) |> json(%{error: "Unknown permissions", names: unknown_permissions})
    end
  end

  def assign_permission(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"id" => id, "permission" => permission}) do
    if InMemoryStore.assign_permission_to_group(id, permission, auth.tenant_id) do
      conn |> put_status(:created) |> json(%{status: "assigned"})
    else
      conn |> put_status(:forbidden) |> json(%{error: "Forbidden"})
    end
  end

  def assign_user(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"id" => id, "user_id" => user_id}) do
    if InMemoryStore.assign_user_to_group(id, user_id, auth.tenant_id) do
      conn |> put_status(:created) |> json(%{status: "assigned"})
    else
      conn |> put_status(:forbidden) |> json(%{error: "Forbidden"})
    end
  end
end
